File: src-tauri/src/ws_server/dispatch_git.rs
//! Git command dispatch for the WebSocket server.
//!
//! Handles all `git_*` commands.

use super::dispatch_helpers::{extract_arg, extract_opt_arg};

/// Dispatch a git command, returning the JSON result.
pub async fn dispatch(
    cmd: &str,
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    match cmd {
        "git_list_mort_branches" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::list_mort_branches(&repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_diff_uncommitted" => {
            let working_directory: String = extract_arg(&args, "workingDirectory")?;
            let result = crate::git_commands::diff_uncommitted(&working_directory).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_fetch" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let remote: Option<String> = extract_opt_arg(&args, "remote");
            crate::git_commands::git_fetch(repo_path, remote).await?;
            Ok(serde_json::Value::Null)
        }
        "git_get_default_branch" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::git_get_default_branch(repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_get_branch_commit" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let branch: String = extract_arg(&args, "branch")?;
            let result =
                crate::git_commands::git_get_branch_commit(repo_path, branch).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_create_branch" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let branch_name: String = extract_arg(&args, "branchName")?;
            let base_branch: String = extract_arg(&args, "baseBranch")?;
            crate::git_commands::git_create_branch(repo_path, branch_name, base_branch)
                .await?;
            Ok(serde_json::Value::Null)
        }
        "git_checkout_branch" => {
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            let branch: String = extract_arg(&args, "branch")?;
            crate::git_commands::git_checkout_branch(worktree_path, branch).await?;
            Ok(serde_json::Value::Null)
        }
        "git_checkout_commit" => {
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            let commit: String = extract_arg(&args, "commit")?;
            crate::git_commands::git_checkout_commit(worktree_path, commit).await?;
            Ok(serde_json::Value::Null)
        }
        "git_delete_branch" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let branch: String = extract_arg(&args, "branch")?;
            crate::git_commands::git_delete_branch(repo_path, branch).await?;
            Ok(serde_json::Value::Null)
        }
        "git_branch_exists" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let branch: String = extract_arg(&args, "branch")?;
            let result =
                crate::git_commands::git_branch_exists(repo_path, branch).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        _ => dispatch_part2(cmd, args).await,
    }
}

/// Second half of git dispatch (split to keep functions under 50 lines).
async fn dispatch_part2(
    cmd: &str,
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    match cmd {
        "git_create_worktree" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            let branch: String = extract_arg(&args, "branch")?;
            crate::git_commands::git_create_worktree(repo_path, worktree_path, branch)
                .await?;
            Ok(serde_json::Value::Null)
        }
        "git_remove_worktree" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let worktree_path: String = extract_arg(&args, "worktreePath")?;
            crate::git_commands::git_remove_worktree(repo_path, worktree_path).await?;
            Ok(serde_json::Value::Null)
        }
        "git_list_worktrees" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::git_list_worktrees(repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_ls_files" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::git_ls_files(repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_ls_files_untracked" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::git_ls_files_untracked(repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_get_head_commit" => {
            let repo_path: String = extract_arg(&args, "repoPath")?;
            let result = crate::git_commands::git_get_head_commit(repo_path).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_diff_files" => dispatch_diff_files(args).await,
        "git_get_branch_commits" => dispatch_branch_commits(args).await,
        "git_diff_commit" => dispatch_diff_commit(args).await,
        "git_diff_range" => dispatch_diff_range(args).await,
        _ => dispatch_part3(cmd, args).await,
    }
}

async fn dispatch_diff_files(
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    let repo_path: String = extract_arg(&args, "repoPath")?;
    let base_commit: String = extract_arg(&args, "baseCommit")?;
    let file_paths: Vec<String> = extract_opt_arg(&args, "filePaths").unwrap_or_default();
    let file_requests: Option<Vec<crate::git_commands::FileDiffRequest>> =
        extract_opt_arg(&args, "fileRequests");
    let result = crate::git_commands::git_diff_files(
        repo_path,
        base_commit,
        file_paths,
        file_requests,
    )
    .await?;
    Ok(serde_json::to_value(result).unwrap())
}

async fn dispatch_branch_commits(
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    let working_directory: String = extract_arg(&args, "workingDirectory")?;
    let branch_name: String = extract_arg(&args, "branchName")?;
    let limit: Option<u32> = extract_opt_arg(&args, "limit");
    let result = crate::git_commands::git_get_branch_commits(
        working_directory,
        branch_name,
        limit,
    )
    .await?;
    Ok(serde_json::to_value(result).unwrap())
}

async fn dispatch_diff_commit(
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    let working_directory: String = extract_arg(&args, "workingDirectory")?;
    let commit_hash: String = extract_arg(&args, "commitHash")?;
    let result =
        crate::git_commands::git_diff_commit(working_directory, commit_hash).await?;
    Ok(serde_json::to_value(result).unwrap())
}

async fn dispatch_diff_range(
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    let working_directory: String = extract_arg(&args, "workingDirectory")?;
    let base_commit: String = extract_arg(&args, "baseCommit")?;
    let result =
        crate::git_commands::git_diff_range(working_directory, base_commit).await?;
    Ok(serde_json::to_value(result).unwrap())
}

/// Third part of git dispatch for remaining commands.
async fn dispatch_part3(
    cmd: &str,
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    match cmd {
        "git_get_merge_base" => {
            let working_directory: String = extract_arg(&args, "workingDirectory")?;
            let branch_a: String = extract_arg(&args, "branchA")?;
            let branch_b: String = extract_arg(&args, "branchB")?;
            let result = crate::git_commands::git_get_merge_base(
                working_directory,
                branch_a,
                branch_b,
            )
            .await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_get_remote_branch_commit" => {
            let working_directory: String = extract_arg(&args, "workingDirectory")?;
            let remote: String = extract_arg(&args, "remote")?;
            let branch: String = extract_arg(&args, "branch")?;
            let result = crate::git_commands::git_get_remote_branch_commit(
                working_directory,
                remote,
                branch,
            )
            .await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_show_file" => {
            let cwd: String = extract_arg(&args, "cwd")?;
            let path: String = extract_arg(&args, "path")?;
            let git_ref: String = extract_arg(&args, "gitRef")?;
            let result =
                crate::git_commands::git_show_file(cwd, path, git_ref).await?;
            Ok(serde_json::to_value(result).unwrap())
        }
        "git_grep" => dispatch_grep(args).await,
        "git_rm" => {
            let working_directory: String = extract_arg(&args, "workingDirectory")?;
            let file_path: String = extract_arg(&args, "filePath")?;
            crate::git_commands::git_rm(working_directory, file_path).await?;
            Ok(serde_json::Value::Null)
        }
        _ => Err(format!("unknown git command: {}", cmd)),
    }
}

async fn dispatch_grep(
    args: serde_json::Value,
) -> Result<serde_json::Value, String> {
    let repo_path: String = extract_arg(&args, "repoPath")?;
    let query: String = extract_arg(&args, "query")?;
    let max_results: Option<u32> = extract_opt_arg(&args, "maxResults");
    let include_patterns: Option<Vec<String>> =
        extract_opt_arg(&args, "includePatterns");
    let exclude_patterns: Option<Vec<String>> =
        extract_opt_arg(&args, "excludePatterns");
    let case_sensitive: Option<bool> = extract_opt_arg(&args, "caseSensitive");
    let result = crate::git_commands::git_grep(
        repo_path,
        query,
        max_results,
        include_patterns,
        exclude_patterns,
        case_sensitive,
    )
    .await?;
    Ok(serde_json::to_value(result).unwrap())
}

File: sidecar/src/dispatch/dispatch-git.ts
/**
 * Git command dispatch.
 * Handles all `git_*` commands.
 */

import { extractArg, extractOptArg } from "../helpers.js";
import { git, gitSafe } from "./git-helpers.js";

export async function dispatchGit(
  cmd: string,
  args: Record<string, unknown>,
): Promise<unknown> {
  switch (cmd) {
    case "git_list_anvil_branches":
      return listAnvilBranches(extractArg(args, "repoPath"));

    case "git_diff_uncommitted":
      return diffUncommitted(extractArg(args, "workingDirectory"));

    case "git_fetch":
      return gitFetch(
        extractArg(args, "repoPath"),
        extractOptArg(args, "remote"),
      );

    case "git_get_default_branch":
      return getDefaultBranch(extractArg(args, "repoPath"));

    case "git_get_branch_commit":
      return git(extractArg(args, "repoPath"), [
        "rev-parse",
        extractArg(args, "branch"),
      ]);

    case "git_init":
      return gitInit(extractArg(args, "path"));

    case "git_create_branch":
      return gitCreateBranch(
        extractArg(args, "repoPath"),
        extractArg(args, "branchName"),
        extractArg(args, "baseBranch"),
      );

    case "git_checkout_branch":
      return git(extractArg(args, "worktreePath"), [
        "checkout",
        extractArg(args, "branch"),
      ]).then(() => null);

    case "git_checkout_commit":
      return git(extractArg(args, "worktreePath"), [
        "checkout",
        "--detach",
        extractArg(args, "commit"),
      ]).then(() => null);

    case "git_delete_branch":
      return git(extractArg(args, "repoPath"), [
        "branch",
        "-D",
        extractArg(args, "branch"),
      ]).then(() => null);

    case "git_branch_exists":
      return gitBranchExists(
        extractArg(args, "repoPath"),
        extractArg(args, "branch"),
      );

    default:
      return dispatchGitPart2(cmd, args);
  }
}

async function dispatchGitPart2(
  cmd: string,
  args: Record<string, unknown>,
): Promise<unknown> {
  switch (cmd) {
    case "git_create_worktree":
      return git(extractArg(args, "repoPath"), [
        "worktree",
        "add",
        "--detach",
        extractArg(args, "worktreePath"),
      ]).then(() => null);

    case "git_remove_worktree":
      return git(extractArg(args, "repoPath"), [
        "worktree",
        "remove",
        "--force",
        extractArg(args, "worktreePath"),
      ]).then(() => null);

    case "git_list_worktrees":
      return listWorktrees(extractArg(args, "repoPath"));

    case "git_ls_files":
      return git(extractArg(args, "repoPath"), ["ls-files"]).then(
        (s) => (s ? s.split("\n") : []),
      );

    case "git_ls_files_untracked":
      return git(extractArg(args, "repoPath"), [
        "ls-files",
        "--others",
        "--exclude-standard",
      ]).then((s) => (s ? s.split("\n") : []));

    case "git_get_head_commit":
      return git(extractArg(args, "repoPath"), ["rev-parse", "HEAD"]);

    case "git_diff_files":
      return gitDiffFiles(args);

    case "git_get_branch_commits":
      return gitGetBranchCommits(args);

    case "git_diff_commit":
      return gitDiffCommit(args);

    case "git_diff_range":
      return gitDiffRange(args);

    default:
      return dispatchGitPart3(cmd, args);
  }
}

async function dispatchGitPart3(
  cmd: string,
  args: Record<string, unknown>,
): Promise<unknown> {
  switch (cmd) {
    case "git_get_merge_base":
      return git(extractArg(args, "workingDirectory"), [
        "merge-base",
        extractArg(args, "branchA"),
        extractArg(args, "branchB"),
      ]);

    case "git_get_remote_branch_commit":
      return git(extractArg(args, "workingDirectory"), [
        "rev-parse",
        `${extractArg<string>(args, "remote")}/${extractArg<string>(args, "branch")}`,
      ]);

    case "git_show_file":
      return git(extractArg(args, "cwd"), [
        "show",
        `${extractArg<string>(args, "gitRef")}:${extractArg<string>(args, "path")}`,
      ]);

    case "git_grep":
      return gitGrep(args);

    case "git_rm":
      return git(extractArg(args, "workingDirectory"), [
        "rm",
        "--force",
        extractArg(args, "filePath"),
      ]).then(() => null);

    default:
      throw new Error(`unknown git command: ${cmd}`);
  }
}

// ── Implementation helpers ─────────────────────────────────────────────

async function listAnvilBranches(repoPath: string): Promise<string[]> {
  const output = await gitSafe(repoPath, ["branch", "--list", "anvil/*"]);
  if (!output) return [];
  return output
    .split("\n")
    .map((b) => b.trim().replace(/^\* /, ""))
    .filter(Boolean);
}

async function diffUncommitted(cwd: string): Promise<string> {
  const tracked = await gitSafe(cwd, ["diff", "HEAD"]);
  const untracked = await gitSafe(cwd, [
    "ls-files",
    "--others",
    "--exclude-standard",
  ]);
  let result = tracked ?? "";

  if (untracked) {
    const { readFile } = await import("node:fs/promises");
    for (const file of untracked.split("\n").filter(Boolean)) {
      try {
        const content = await readFile(`${cwd}/${file}`, "utf-8");
        const lines = content.split("\n");
        const patch = lines.map((l) => `+${l}`).join("\n");
        result += `\ndiff --git a/${file} b/${file}\nnew file mode 100644\n--- /dev/null\n+++ b/${file}\n@@ -0,0 +1,${lines.length} @@\n${patch}\n`;
      } catch {
        // Skip binary or unreadable files
      }
    }
  }
  return result;
}

async function gitFetch(
  repoPath: string,
  remote?: string,
): Promise<null> {
  await git(repoPath, ["fetch", remote ?? "origin"]);
  return null;
}

async function getDefaultBranch(repoPath: string): Promise<string> {
  // Strategy 1: refs/remotes/origin/HEAD
  const ref = await gitSafe(repoPath, [
    "symbolic-ref",
    "refs/remotes/origin/HEAD",
  ]);
  if (ref) {
    const branch = ref.replace("refs/remotes/origin/", "");
    if (branch && branch !== ref) return branch;
  }

  // Strategy 2: git config init.defaultBranch
  const configured = await gitSafe(repoPath, [
    "config",
    "init.defaultBranch",
  ]);
  if (configured) return configured;

  // Strategy 3: Test common names
  for (const name of ["main", "master", "develop", "trunk"]) {
    const exists = await gitBranchExists(repoPath, name);
    if (exists) return name;
  }

  // Strategy 4: current branch
  const current = await gitSafe(repoPath, [
    "branch",
    "--show-current",
  ]);
  if (current) return current;

  return "main";
}

async function gitInit(path: string): Promise<null> {
  const { mkdir } = await import("node:fs/promises");
  await mkdir(path, { recursive: true });
  await git(path, ["init"]);
  return null;
}

async function gitCreateBranch(
  repoPath: string,
  branchName: string,
  baseBranch: string,
): Promise<null> {
  const exists = await gitBranchExists(repoPath, branchName);
  if (exists) throw new Error(`Branch already exists: ${branchName}`);
  await git(repoPath, ["branch", branchName, baseBranch]);
  return null;
}

async function gitBranchExists(
  repoPath: string,
  branch: string,
): Promise<boolean> {
  const result = await gitSafe(repoPath, [
    "show-ref",
    "--verify",
    "--quiet",
    `refs/heads/${branch}`,
  ]);
  return result !== null;
}

interface WorktreeInfo {
  path: string;
  branch: string | null;
  isBare: boolean;
}

async function listWorktrees(repoPath: string): Promise<WorktreeInfo[]> {
  const output = await git(repoPath, ["worktree", "list", "--porcelain"]);
  const worktrees: WorktreeInfo[] = [];
  let current: Partial<WorktreeInfo> = {};

  for (const line of output.split("\n")) {
    if (line.startsWith("worktree ")) {
      if (current.path) worktrees.push(current as WorktreeInfo);
      current = { path: line.slice(9), branch: null, isBare: false };
    } else if (line.startsWith("branch ")) {
      current.branch = line.slice(7).replace("refs/heads/", "");
    } else if (line === "bare") {
      current.isBare = true;
    } else if (line === "" && current.path) {
      worktrees.push(current as WorktreeInfo);
      current = {};
    }
  }
  if (current.path) worktrees.push(current as WorktreeInfo);
  return worktrees;
}

async function gitDiffFiles(
  args: Record<string, unknown>,
): Promise<string> {
  const repoPath = extractArg<string>(args, "repoPath");
  const baseCommit = extractArg<string>(args, "baseCommit");
  const filePaths =
    extractOptArg<string[]>(args, "filePaths") ?? [];

  const gitArgs = ["diff", baseCommit, "--"];
  gitArgs.push(...filePaths);
  return git(repoPath, gitArgs);
}

interface GitCommit {
  hash: string;
  shortHash: string;
  message: string;
  author: string;
  authorEmail: string;
  date: string;
  relativeDate: string;
}

async function gitGetBranchCommits(
  args: Record<string, unknown>,
): Promise<GitCommit[]> {
  const cwd = extractArg<string>(args, "workingDirectory");
  const branch = extractArg<string>(args, "branchName");
  const limit = extractOptArg<number>(args, "limit") ?? 50;

  const format = "%H%x00%h%x00%s%x00%an%x00%ae%x00%aI%x00%ar";
  const output = await git(cwd, [
    "log",
    "--first-parent",
    `--format=${format}`,
    branch,
    `-n`,
    String(limit),
  ]);

  if (!output) return [];
  return output.split("\n").map((line) => {
    const [hash, shortHash, message, author, authorEmail, date, relativeDate] =
      line.split("\0");
    return { hash, shortHash, message, author, authorEmail, date, relativeDate };
  });
}

async function gitDiffCommit(
  args: Record<string, unknown>,
): Promise<string> {
  const cwd = extractArg<string>(args, "workingDirectory");
  const commitHash = extractArg<string>(args, "commitHash");

  // Try parent diff first
  const result = await gitSafe(cwd, [
    "diff",
    `${commitHash}^..${commitHash}`,
  ]);
  if (result !== null) return result;

  // Root commit fallback
  return git(cwd, ["show", "--format=", commitHash]);
}

async function gitDiffRange(
  args: Record<string, unknown>,
): Promise<string> {
  const cwd = extractArg<string>(args, "workingDirectory");
  const baseCommit = extractArg<string>(args, "baseCommit");
  return git(cwd, ["diff", baseCommit]);
}

interface GrepMatch {
  filePath: string;
  lineNumber: number;
  lineContent: string;
}

async function gitGrep(
  args: Record<string, unknown>,
): Promise<{ matches: GrepMatch[]; truncated: boolean }> {
  const repoPath = extractArg<string>(args, "repoPath");
  const query = extractArg<string>(args, "query");
  const maxResults = extractOptArg<number>(args, "maxResults") ?? 5000;
  const caseSensitive =
    extractOptArg<boolean>(args, "caseSensitive") ?? true;

  const gitArgs = ["grep", "-n", "-F"];
  if (!caseSensitive) gitArgs.push("-i");
  gitArgs.push(query);

  const output = await gitSafe(repoPath, gitArgs);
  if (!output) return { matches: [], truncated: false };

  const lines = output.split("\n").filter(Boolean);
  const truncated = lines.length > maxResults;
  const matches = lines.slice(0, maxResults).map((line) => {
    const firstColon = line.indexOf(":");
    const secondColon = line.indexOf(":", firstColon + 1);
    return {
      filePath: line.slice(0, firstColon),
      lineNumber: parseInt(line.slice(firstColon + 1, secondColon), 10),
      lineContent: line.slice(secondColon + 1),
    };
  });

  return { matches, truncated };
}
